hail/python/hailtop/batch/batch_pool_executor.py

```diff
@@ -240,23 +240,47 @@ async def async_map(self,
                 chunk for chunk in iterables_chunks if len(chunk) > 0]
             fn = chunk(fn)
             iterables = iterables_chunks
-        submissions = [self.async_submit(fn, *arguments)
-                       for arguments in zip(*iterables)]
-        futures: List[BatchPoolFuture] = await asyncio.gather(*submissions)
+
+        tasks = [asyncio.create_task(self.async_submit(fn, *arguments))
+                 for arguments in zip(*iterables)]
+
+        try:
+            bp_futures: List[BatchPoolFuture] = await asyncio.gather(*tasks)
+        finally:
+            _, exc, _ = sys.exc_info()
+            if exc is not None:
+                for task in tasks:
+                    if not task.done():
+                        task.cancel()
+
+                if tasks:
+                    await asyncio.wait(tasks)
+
+                bp_futures = []
+                for task in tasks:
+                    assert task.done()
+                    if not task.cancelled() and not task.exception():
+                        bp_future = task.result()
+                        bp_futures.append(bp_future)
+
+                if bp_futures:
+                    await asyncio.wait([bp_fut.async_cancel() for bp_fut in bp_futures])
```

Comment:
I was following the structure we used in `bounded_gather2_raise_exceptions` in hailtop.utils. I switched them both.

Reply:
Ahh, I understand. We need to re-raise the exception, which means that we end up with the same exception twice (because we're re-raising the exception in the handler for that exception).

OK, let's go back to the finally, if exc is not none approach. My bad! Sorry!